File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
class RandomizedSet {
    List<Integer> list = new LinkedList<>();
    Map<Integer, Integer> map = new HashMap<>();
    Random rnd = new Random();
    /** Initialize your data structure here. */
    public RandomizedSet() { }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        if (!map.containsKey(val)) {
            list.add(val);
            map.put(val, list.size() - 1);
            return true;
        }
        return false;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        if (!map.containsKey(val)) return false;
        int index = map.get(val); 
        swap(index, list.size() - 1);
        map.put(list.get(index), index);
        map.remove(val);
        list.remove(list.size() - 1);
        return true;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
        int index = rnd.nextInt(list.size());
        return list.get(index);
    }
    
    private void swap(int a, int b) {
        int temp = list.get(a);
        list.set(a, list.get(b));
        list.set(b, temp);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.js

// var RandomizedSet = function() {
    
// };

// /** 
//  * @param {number} val
//  * @return {boolean}
//  */
// RandomizedSet.prototype.insert = function(val) {
    
// };

// /** 
//  * @param {number} val
//  * @return {boolean}
//  */
// RandomizedSet.prototype.remove = function(val) {
    
// };

// /**
//  * @return {number}
//  */
// RandomizedSet.prototype.getRandom = function() {
    
// };

/** 
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */
class RandomizedSet {
  /**
   * Initialize your data structure here.
   */
  constructor() {
    this.nums = [];
    this.pos = {};
  }

  /**
   * Inserts a value to the set. Returns true if the set did not already contain the specified element.
   * @param {number} val
   * @return {boolean}
   */
  insert(val) {
    if (this.pos[val] != null) return false;

    this.nums.push(val);
    this.pos[val] = this.nums.length - 1;
    return true;
  }

  /**
   * Removes a value from the set. Returns true if the set contained the specified element.
   * @param {number} val
   * @return {boolean}
   */
  remove(val) {
    if (this.pos[val] == null) return false;

    const i = this.pos[val];
    const lastNum = this.nums[this.nums.length - 1];

    // Swap the last num with val
    this.nums[i] = lastNum;
    this.pos[lastNum] = i;

    // Remove val
    this.nums.pop();
    delete this.pos[val];

    return true;
  }

  /**
   * Get a random element from the set.
   * @return {number}
   */
  getRandom() {
    const i = ~~(Math.random() * this.nums.length);
    return this.nums[i]
  }
}